[sw, tock] Extend downstream tock development environment

silvestrst opened the pull request:
This is my take on the downstream Tock development environment and request for comments. Please let me know what you think of this approach.

When it comes to actual PR, some of the commits might be merged or split out further, but for now this should suffice.

For simplicity I have added git hash to external dependencies, otherwise some of our existing `board/opentitan` code is behind the upstream, and Tock `kernel` crate API has changed.

I understand that `cf6191f` might be slightly controversial, however this the cleanest way I found to keep Jon Flatley's "use local repo" change without creating 3x pairs of `Cargo_local.toml` and `Cargo_remote.toml`. I quite like this approach as it clearly isolates the external dependencies from our version controlled source. This change touches a lot of files, but the changes are very minimalistic - and let us keep using namespaces through the code as they are upstream.

TODO:

- [x] Update .gitignore to include/exclude relevant `.lock` and `.toml` files.

- [ ] I don't think there is any point in `ealgrey/source/<uart.rs/gpio.rs>` `earlgrey` should always be instantiated in one way (base addresses, etc...). I think `UART0`, `GPIO0` should be moved to `chip.rs` or even `lib.rs`.

-  There is no reason why `plic.rs` and `rv_timer.rs` shouldn't live in `peripherals` instead of `earlgrey`. If there is need for some extra wrapping, `earlgrey` can add wrapping on top of `peripherals/plic.rs` (Non urgent, subject of further follow-up PRs).

- [x] Need to add device specific configuration (conditional compilation), same as in https://github.com/tock/tock/pull/1741
Planning to update the upstream PR, and implement the downstream device config in the same way.

- sync up with upstream and update `main.rs` and other out of date source (Non urgent, subject of further follow-up PRs).

- [x] Add Jon Flatley's `bindgen` change.

- [ ] Add my linking against DIF libraries change.

P.S. there will be a lot of linter issues, I haven't ran it through `clang-format` yet.

lenary commented on sw/device/tock/boards/opentitan/tock_dependencies/Cargo.lock:
```diff
@@ -0,0 +1 @@
+/home/svt/development/repos/opentitan/sw/device/tock/boards/opentitan/tock_dependencies/Cargo_remote.lock
```
You need to update the `.gitignore` to ignore this file.

silvestrst replied:
> I don't think we should be nesting cargo packages like you have with boards/opentitan/earlgrey and boards/opentitan/earlgrey/peripherals (and also boards/opentitan/tock_dependencies).

We don't have to nest them, I am happy to pull it out outside `board/` directory, but could you please why we shouldn't do it in your opinion in more detail.

> How are sw/device/tock/boards/opentitan/src/*.rs built? Don't they need a Cargo.toml?

They do. It is the `Cargo.toml` in the `boards/opentitan` directory.

silvestrst replied:
> They do. It is the `Cargo.toml` in the `boards/opentitan` directory. However, I couldn't find it in the committed files, I think it's maybe to do with the fact that the symlink with this name was already in git. I will investigate this.

It is because `Cargo.toml`used to be a symlink and was in `.gitignore`. Nice catch, thank you!

lenary commented on sw/device/tock/boards/opentitan/tock_dependencies/Cargo.toml:
```diff
@@ -0,0 +1 @@
+/home/svt/development/repos/opentitan/sw/device/tock/boards/opentitan/tock_dependencies/Cargo_remote.toml
```
You need to update the `.gitignore` to ignore this file.

silvestrst replied:
Good point, thanks.

silvestrst replied:
Thanks for mentioning this.

jon-flatley replied:
`Cargo_remote.lock` should **not** be ignored as this is how we pin the build to upstream tock revisions. In general the `.gitignore` should be updated to reflect the new paths.

silvestrst commented:
In the latest push:
* Modified `.gitignore`.
* Fixed up the last commit in series to include `board/opentitan/Cargo.<toml/lock>`.

jon-flatley commented on sw/device/tock/boards/opentitan/earlgrey/peripherals/src/lib.rs:
```diff
@@ -0,0 +1,11 @@
+//! Implementations for generic LowRISC peripherals.
+
+#![feature(asm, concat_idents, const_fn, core_intrinsics)]
+#![feature(in_band_lifetimes)]
+#![feature(exclusive_range_pattern)]
```
Inconsistent pattern here, can this be made into a single `#![feature(...)]` or each broken out into distinct attributes?

silvestrst replied:
I think this probably should be done as a follow-up. At this point it has simply been copied from the tock repository of the revision that matches our existing downstream tock board implementation.

jon-flatley commented on sw/device/tock/boards/opentitan/earlgrey/src/lib.rs:
```diff
@@ -0,0 +1,15 @@
+//! Drivers and chip support for the Ibex soft core.
+
+#![feature(asm, concat_idents, const_fn, naked_functions)]
+#![feature(in_band_lifetimes)]
```
Same as above, can these be combined?

silvestrst replied:
Please see the comment above.

jon-flatley commented on sw/device/tock/meson.build:
```diff
@@ -43,7 +47,7 @@ tock_raw = custom_target(
   output: target,
   console: true,
   build_always_stale: true,
-  build_by_default: false,
+  build_by_default: true,
```
We shouldn't enable building Rust targets by default until we have toolchain requirements sorted out at the OT project level. This will break the build for those without rustup, including CI builds.

silvestrst replied:
Oh, absolutely. This is just a temporary change for work I am doing. I am not intending to merge this.

jon-flatley commented on sw/device/tock/boards/opentitan/tock_dependencies/Cargo_remote.toml:
```diff
@@ -0,0 +1,12 @@
+edition = "2018"
+
+[dependencies]
+components = { git = "https://github.com/tock/tock", rev = "d37ef3d2" }
+rv32i = { git = "https://github.com/tock/tock", rev = "d37ef3d2" }
+capsules = { git = "https://github.com/tock/tock", rev = "d37ef3d2"}
+kernel = { git = "https://github.com/tock/tock", rev = "d37ef3d2" }
+tock_rt0 = { git = "https://github.com/tock/tock", rev = "d37ef3d2" }
```
I don't think we want to specify the revision here, it's handled in `Cargo.lock`. Eventually we will need CI tooling to advance this revision to upstream ToT, which I believe Cargo can mostly do for us if we use `Cargo.lock`.

silvestrst replied:
Yup I agree, I am not sure if I use "pre this change" Cargo.lock, whether it will keep the revision that it used to have, I can try it later.

Just to explain why I forced the revision here - it's just to make sure that we use the same revision as before this change for now. As the upstream kernel interface changed slightly, so latest tock version breaks our downstream board.

silvestrst commented:
@jon-flatley thank you for your review, I will go through your comments more thoroughly tomorrow.

silvestrst commented on sw/device/tock/boards/opentitan/Cargo.toml:
```diff
@@ -18,9 +18,5 @@ opt-level = "z"
-ibex = { git = "https://github.com/tock/tock" }
-lowrisc = { git = "https://github.com/tock/tock" }
+earlgrey = { path = "./earlgrey" }
+tock_dependencies = { path = "./tock_dependencies" } 
```
Sure, will do.

silvestrst commented:
@lenary reshuffled the directories, could you please have a look?

imphil commented on sw/device/tock/boards/opentitan/Cargo.toml:
```diff
@@ -18,9 +18,5 @@ opt-level = "z"
-ibex = { git = "https://github.com/tock/tock" }
-lowrisc = { git = "https://github.com/tock/tock" }
+earlgrey = { path = "../../chips/earlgrey" }
+tock_deps = { path = "../../tock_deps" }
```
missing newline.

silvestrst replied:
Does cargo need a newline at the EOF as well?

silvestrst commented:
* Updated description of the first commit (forgot to do after last push, and the directories layout has changed).
* Renamed `chips` directory to `chip`.

lenary commented:
I am happy with the directory structure (mostly I am happy to defer to @jon-flatley and @gkelly on that question, as they're more familiar with Cargo and usual rust conventions than I am).

silvestrst commented:
Made it possible to compile for fpga or verilator from command line (through `TOCK_FEATURES` environment variable).

jon-flatley commented:
> Made it possible to compile for fpga or verilator from command line (through `TOCK_FEATURES` environment variable).

Do we want to use an environment variable for this or create two separate targets based on https://github.com/lowRISC/opentitan/tree/master/sw/device/lib/arch?

lenary commented:
@jon-flatley that sounds good, but it would be good if we could do that in a follow-up PR, as I don't think it should hold up DIF integration (which is blocked by this PR).

The questions to be answered in this PR are:
- [ ] Does this directory structure make sense? 
    I think it does, though there's an open question about calling the intermediate package `tock_deps`.
- [ ] Is this the minimum we need to do to start downstream tock development? 
    I believe it is, but I need someone else to confirm this.

silvestrst commented:
> Do we want to use an environment variable for this or create two separate targets based on

Creating separate targets would be nicer, but that is something that requires more consideration and could be done in future (providing it's not too much effort). For now, I think environment variable is okay (easiest option).

silvestrst commented:
I am piling commits, but when it comes to non-draft PR, I probably squash these commits in two/three separate PRs.

As @lenary said - the most important thing here is to decide whether the layout is good enough or not. From there we can start moving forward.

jon-flatley commented:
> @jon-flatley that sounds good, but it would be good if we could do that in a follow-up PR, as I don't think it should hold up DIF integration (which is blocked by this PR).
> 
> The questions to be answered in this PR are:
> 
> * [ ]  Does this directory structure make sense?
>   I think it does, though there's an open question about calling the intermediate package `tock_deps`.
> * [ ]  Is this the minimum we need to do to start downstream tock development?
>   I believe it is, but I need someone else to confirm this.

Fair enough, we can move the meson targets around later without too much trouble.

I think the `_deps` is redundant, if it doesn't cause collisions I prefer to just call the crate `tock`.

Otherwise LGTM, I think the directory structure makes sense.

silvestrst commented:
Closed in favour of #2119 , expect other related PRs.
